src/system/sys_profiler.cc

```diff
@@ -630,21 +829,143 @@ void Profiler::uninitialize() {
         ASSERT( _instance->actual_node==0 , "Forbidden to uninitialize the Profiler when actual timer is not zero (but '%s').\n",
                 _instance->timers_[_instance->actual_node].tag().c_str());
         _instance->stop_timer(0);
+        set_memory_monitoring(false);
         delete _instance;
         _instance = NULL;
     }
 }
 
+bool Profiler::monitor_memory = false;
+void Profiler::set_memory_monitoring(const bool value) {
+    if (value && !monitor_memory) {
+        cout << "Memory monitoring is ON" << endl;
+    }
+    if (!value && monitor_memory) {
+        cout << "Memory monitoring is OFF" << endl;
+    }
+    monitor_memory = value;
+}
+
+bool Profiler::get_memory_monitoring() {
+    return monitor_memory;
+}
+
+unordered_map_with_alloc & MemoryAlloc::malloc_map() {
+    static unordered_map_with_alloc static_malloc_map;
+    return static_malloc_map;
+}
+
+unordered_map_with_alloc & MemoryAlloc::init_overhead_map() {
+    static unordered_map_with_alloc static_init_overhead_map;
+    return static_init_overhead_map;
+}
+
+int MemoryAlloc::sum(unordered_map_with_alloc & map) {
+    int total = 0;
+    for (auto it=map.begin(); it!=map.end(); it++)
+        total += it->second;
+    return total;
+}
+
+void * Profiler::operator new (size_t size) {
+    return malloc (size);
+}
+
+
+void *operator new (std::size_t size) OPERATOR_NEW_THROW_EXCEPTION {
+	void * p = malloc(size);
+    
+    if (!Profiler::is_initialized()) {
+        MemoryAlloc::init_overhead_map()[(long)p] = static_cast<int>(size);
+    }else
+    
+    if (Profiler::get_memory_monitoring()) {
+        Profiler::instance()->notify_malloc(size);
+        MemoryAlloc::malloc_map()[(long)p] = static_cast<int>(size);
+    }
+
+	return p;
+}
+
+void *operator new[] (std::size_t size) OPERATOR_NEW_THROW_EXCEPTION {
+    void * p = malloc(size);
+    
+    if (!Profiler::is_initialized()) {
+        MemoryAlloc::init_overhead_map()[(long)p] = static_cast<int>(size);
+    }else
+    
+    if (Profiler::get_memory_monitoring()) {
+        Profiler::instance()->notify_malloc(size);
+        MemoryAlloc::malloc_map()[(long)p] = static_cast<int>(size);
+    }
+    
+
+	return p;
+}
+
+void *operator new[] (std::size_t size, const std::nothrow_t&) throw() {
+    void * p = malloc(size);
+    
+    if (!Profiler::is_initialized()) {
+        MemoryAlloc::init_overhead_map()[(long)p] = static_cast<int>(size);
+    }else
+    
+    if (Profiler::get_memory_monitoring()) {
+        Profiler::instance()->notify_malloc(size);
+        MemoryAlloc::malloc_map()[(long)p] = static_cast<int>(size);
+    }
+
+	return p;
+}
+
+void operator delete( void *p) throw() {
+    if (Profiler::get_memory_monitoring()) {
```

Comment:
Similar to new operators call Profiler::notify_free, with the pointer as parameter and do the whole stuff there. This also reduce code redundancy.